app/controllers/course/assessment/question/programming_controller.rb

```diff
@@ -58,4 +44,31 @@ def programming_question_params
       skill_ids: []
     )
   end
+
+  def save_and_redirect(template)
+    if @programming_question.save
+      if @programming_question.import_job
+        redirect_to job_path(@programming_question.import_job)
+      else
+        redirect_to course_assessment_path(current_course, @assessment),
+                    success: t('.success')
+      end
+    else
+      render template
+    end
+  end
+
+  def save_and_render_json
+    if @programming_question.save! && @programming_question.import_job
+      @redirect_url = job_path(@programming_question.import_job)
+    end
+
+    render '_props'
+  end
+
+  def programming_package_service(params = nil)
```

Comment:
👍  cache the service then ? `@service ||=` I think you don't want to create a new instance every time we call it.